# Ordered Logistic Regression

## 1. Data Preparation

### 1.1) Import Libraries & Dataset

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf

In [2]:
#Import data
NHL_game = pd.read_csv("../../Data/Week 1/NHL_game2.csv")
#Display data
display(NHL_game[0:10])

comp_id         competition_name        date   gid  goals_against  \
0        2  2017 NHL Regular Season  2018-03-20  2254            3.0   
1        2  2017 NHL Regular Season  2018-03-07  2050            4.0   
2        2  2017 NHL Regular Season  2017-12-06   844            2.0   
3        2  2017 NHL Regular Season  2017-11-23   659            2.0   
4        2  2017 NHL Regular Season  2018-03-13  2147            7.0   
5        2  2017 NHL Regular Season  2017-11-04   407            4.0   
6        2  2017 NHL Regular Season  2018-04-01  2439            4.0   
7        2  2017 NHL Regular Season  2018-02-18  1797            4.0   
8        2  2017 NHL Regular Season  2017-10-09    71            3.0   
9        2  2017 NHL Regular Season  2017-11-17   567            1.0   

   goals_for  hgd home_away            team_name  tid tricode  type  win  \
0        4.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
1        5.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
2        6.0  4.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
3        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
4        4.0 -3.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
5        5.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6        1.0 -3.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
7        3.0 -1.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
8        4.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
9        6.0  5.0      home  Tampa Bay Lightning   25     TBL     2  1.0   

   win.ord  year  
0      2.0  2017  
1      2.0  2017  
2      2.0  2017  
3      2.0  2017  
4      0.0  2017  
5      2.0  2017  
6      0.0  2017  
7      0.0  2017  
8      2.0  2017  
9      2.0  2017

### 1.2) Extract 2016 regular season records, named "NHL_reg_2016" and then explore the resulting dataset

In [3]:
# Filter rows for 2016 NHL Regular Season
NHL_reg_2016 =  NHL_game[NHL_game.competition_name == '2016 NHL Regular Season']

In [4]:
# Display top 10 rows
display(NHL_reg_2016[0:10])

comp_id         competition_name        date   gid  goals_against  \
6395     2734  2016 NHL Regular Season  2017-04-01  5038            2.0   
6396     2734  2016 NHL Regular Season  2016-11-05  3068            1.0   
6397     2734  2016 NHL Regular Season  2017-03-12  4719            2.0   
6398     2734  2016 NHL Regular Season  2017-01-13  3987            2.0   
6399     2734  2016 NHL Regular Season  2017-03-21  4876            5.0   
6400     2734  2016 NHL Regular Season  2017-03-18  4830            5.0   
6401     2734  2016 NHL Regular Season  2017-03-30  5008            3.0   
6402     2734  2016 NHL Regular Season  2017-02-28  4555            1.0   
6403     2734  2016 NHL Regular Season  2017-02-05  4271            4.0   
6404     2734  2016 NHL Regular Season  2016-12-04  3462            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
6395        1.0 -1.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
6396        4.0  3.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6397        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6398        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6399        3.0 -2.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
6400        3.0 -2.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
6401        5.0  2.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6402        5.0  4.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6403        5.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6404        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   

      win.ord  year  
6395      0.0  2016  
6396      2.0  2016  
6397      2.0  2016  
6398      2.0  2016  
6399      0.0  2016  
6400      0.0  2016  
6401      2.0  2016  
6402      2.0  2016  
6403      2.0  2016  
6404      2.0  2016

In [5]:
#Examine the size of data
NHL_reg_2016.shape

(2424, 15)

In [6]:
#Explore the types of variables (data) and missing values in a data frame
NHL_reg_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2424 entries, 6395 to 16859
Data columns (total 15 columns):
comp_id             2424 non-null int64
competition_name    2424 non-null object
date                2424 non-null object
gid                 2424 non-null int64
goals_against       2424 non-null float64
goals_for           2424 non-null float64
hgd                 2424 non-null float64
home_away           2424 non-null object
team_name           2424 non-null object
tid                 2424 non-null int64
tricode             2424 non-null object
type                2424 non-null int64
win                 2424 non-null float64
win.ord             2424 non-null float64
year                2424 non-null int64
dtypes: float64(5), int64(5), object(5)
memory usage: 303.0+ KB


In [7]:
#Obtain basic descriptive statistics(round off the statistics properly)
NHL_reg_2016.describe().round(decimals = 3).round({'comp_id':0, 'gid':0, 'tid':0, 'year':0})

comp_id     gid  goals_against  goals_for       hgd     tid    type  \
count   2424.0  2424.0       2424.000   2424.000  2424.000  2424.0  2424.0   
mean    2734.0  3958.0          2.797      2.797     0.320    25.0     2.0   
std        0.0   702.0          1.641      1.641     2.361    18.0     0.0   
min     2734.0  2734.0          0.000      0.000    -7.000     1.0     2.0   
25%     2734.0  3353.0          2.000      2.000    -1.000    10.0     2.0   
50%     2734.0  3961.0          3.000      3.000     1.000    20.0     2.0   
75%     2734.0  4560.0          4.000      4.000     2.000    41.0     2.0   
max     2734.0  5179.0         10.000     10.000    10.000    66.0     2.0   

            win   win.ord    year  
count  2424.000  2424.000  2424.0  
mean      0.500     1.000  2016.0  
std       0.498     0.996     0.0  
min       0.000     0.000  2016.0  
25%       0.000     0.000  2016.0  
50%       0.500     1.000  2016.0  
75%       1.000     2.000  2016.0  
max       1.000     2.000  2016.0

### 1.3) Create home dummy variable & Merge it into the current dataset

In [8]:
#Home dummy variable
home_dummy=pd.get_dummies(NHL_reg_2016['home_away'])
home_dummy.head()

away  home
6395     0     1
6396     0     1
6397     0     1
6398     0     1
6399     0     1

In [9]:
# Merge Home dummy variable into NHL dataset
NHL_reg_2016 = NHL_reg_2016.merge(home_dummy, left_index=True, right_index=True)
NHL_reg_2016.head()

comp_id         competition_name        date   gid  goals_against  \
6395     2734  2016 NHL Regular Season  2017-04-01  5038            2.0   
6396     2734  2016 NHL Regular Season  2016-11-05  3068            1.0   
6397     2734  2016 NHL Regular Season  2017-03-12  4719            2.0   
6398     2734  2016 NHL Regular Season  2017-01-13  3987            2.0   
6399     2734  2016 NHL Regular Season  2017-03-21  4876            5.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
6395        1.0 -1.0      home  Tampa Bay Lightning   25     TBL     2  0.0   
6396        4.0  3.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6397        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6398        3.0  1.0      home  Tampa Bay Lightning   25     TBL     2  1.0   
6399        3.0 -2.0      home  Tampa Bay Lightning   25     TBL     2  0.0   

      win.ord  year  away  home  
6395      0.0  2016     0     1  
6396      2.0  2016     0     1  
6397      2.0  2016     0     1  
6398      2.0  2016     0     1  
6399      0.0  2016     0     1

## Self Test: Add pythagorean winning % to the current dataset
### Add Pythagorean Winning percentage
1) Sort the dataframe by game in order (i.e., gid)
    (Note you will need to reset your index after you sort the NHL_reg_2016 data)

2) Calculate cumulative GF and GA for each team

3) Calculate Pythagorean win-percent

In [10]:
#1) Sort the dataframe by game in order (i.e., gid)
#2) Calculate cumulative GF and GA for each team

#Your Code Here

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
0           5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1           4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3           5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4           4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
...         ...  ...       ...                  ...  ...     ...   ...  ...   
2419        0.0 -2.0      home  Washington Capitals    5     WSH     2  0.0   
2420        3.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
2421        4.0  1.0      home        Anaheim Ducks   21     ANA     2  1.0   
2422        2.0  3.0      away    Vancouver Canucks   53     VAN     2  0.0   
2423        5.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   

      win.ord  year  away  home  cumGF  cumGA  
0         2.0  2016     0     1    5.0    4.0  
1         0.0  2016     1     0    4.0    5.0  
2         0.0  2016     0     1    2.0    5.0  
3         2.0  2016     1     0    5.0    2.0  
4         0.0  2016     1     0    4.0    7.0  
...       ...   ...   ...   ...    ...    ...  
2419      0.0  2016     0     1  265.0  186.0  
2420      0.0  2016     1     0  197.0  204.0  
2421      2.0  2016     0     1  233.0  210.0  
2422      0.0  2016     1     0  178.0  238.0  
2423      2.0  2016     0     1  245.0  211.0  

[2424 rows x 19 columns]

In [11]:
# Calculate Pythagorean win-percent

#Your Code Here

comp_id         competition_name        date   gid  goals_against  \
0     2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1     2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2     2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3     2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4     2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
5     2734  2016 NHL Regular Season  2016-10-13  2738            4.0   
6     2734  2016 NHL Regular Season  2016-10-13  2740            2.0   
7     2734  2016 NHL Regular Season  2016-10-13  2740            1.0   
8     2734  2016 NHL Regular Season  2016-10-13  2742            1.0   
9     2734  2016 NHL Regular Season  2016-10-13  2742            4.0   

   goals_for  hgd home_away            team_name  tid tricode  type  win  \
0        5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1        4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2        2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3        5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4        4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
5        7.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   
6        1.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
7        2.0  1.0      home      San Jose Sharks   16     SJS     2  1.0   
8        4.0 -3.0      away   Montréal Canadiens    2     MTL     2  1.0   
9        1.0 -3.0      home       Buffalo Sabres   17     BUF     2  0.0   

   win.ord  year  away  home  cumGF  cumGA  pyth_wpct  
0      2.0  2016     0     1    5.0    4.0   0.609756  
1      0.0  2016     1     0    4.0    5.0   0.390244  
2      0.0  2016     0     1    2.0    5.0   0.137931  
3      2.0  2016     1     0    5.0    2.0   0.862069  
4      0.0  2016     1     0    4.0    7.0   0.246154  
5      2.0  2016     0     1    7.0    4.0   0.753846  
6      0.0  2016     1     0    1.0    2.0   0.200000  
7      2.0  2016     0     1    2.0    1.0   0.800000  
8      2.0  2016     1     0    4.0    1.0   0.941176  
9      0.0  2016     0     1    1.0    4.0   0.058824

In [12]:
NHL_reg_2016.shape

(2424, 20)

### 2. Ordered Logistic Regression

#### To run ordered logit regression in Python, we need to install a new library, "bevel" which provide codes to fit the ordered logit model and obtain the thresholds of each categorical outcome and the coefficient for the IV. 

### 2.1) Import libraries and fit the ordered logit model

In [13]:
#Import bevel library to fit ordit model
from bevel.linear_ordinal_regression import OrderedLogit
import pandas as pd
import numpy as np

In [14]:
# Fit the ordered logit model and print the result
ol = OrderedLogit()
ol.fit(NHL_reg_2016['pyth_wpct'], NHL_reg_2016['win.ord'])
ol.print_summary()

n=2424
                  beta  se(beta)      p  lower 0.95  upper 0.95     
attribute names                                                     
column_1        5.5758    0.4289 0.0000      4.7351      6.4165  ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Somers' D = 0.217


### 2.2 ) Obtain the coefficients, the intercepts, and the standard errors
The intercepts define the thresholds between a loss and a draw (L|D) and draw and a win (D|W). The probabilities are dependent on (a) the coefficient (i.e., pyth_wpct) and (b) the intercepts. 

In [15]:
# Obtain the coefficients and the intercepts
print(f'beta = {ol.coef_[0]:.4f}')
print(f'interceptL|D = {ol.coef_[1]:.4f}')
print(f'interceptD|W = {ol.coef_[2]:.4f}')
print(ol.coef_)

beta = 5.5758
interceptL|D = 2.7593
interceptD|W = 2.7985
[5.5757812  2.75932539 2.79854712]


In [16]:
# Standard errors
print(ol.se_)

[0.428934   0.21743409 0.2176353 ]


### 2.3) Obtain the fitted values from the model
We can obtain 1) linear product of the ordered regression model, 2) fitted proabilities of being in each outcome of the game and 3) fitted game results from the probabilities. 

In [17]:
# Obtain the linear product from regression
ol.predict_linear_product(NHL_reg_2016[['pyth_wpct']])

array([[3.39986658],
       [2.17591461],
       [0.76907327],
       ...,
       [3.07659988],
       [2.00007983],
       [3.20133019]])

In [18]:
# Obtain the fitted probabilities of each event 
ol.predict_probabilities(NHL_reg_2016[['pyth_wpct']])

array([[0.34512421, 0.00891767, 0.64595812],
       [0.64185185, 0.00896519, 0.34918297],
       [0.87976981, 0.00408727, 0.11614292],
       ...,
       [0.42134012, 0.00959111, 0.56906877],
       [0.68118991, 0.00845661, 0.31035348],
       [0.39126337, 0.00938049, 0.59935614]])

In [19]:
# Returns the output with highest probability 
ol.predict_class(NHL_reg_2016[['pyth_wpct']])

array([2, 0, 0, ..., 2, 0, 2])

In [20]:
#Convert array into dataframe to compare with the true values
fitted_results = ol.predict_class(NHL_reg_2016[['pyth_wpct']])
fitted_results_df = pd.DataFrame(fitted_results, columns = ['fitted_results'])
fitted_results_df.shape

(2424, 1)

In [21]:
# Append the fitted results to NHL_reg_2016 data frame to compare them with true values in win.ord column
NHL_reg_2016['fitted_results'] = fitted_results_df

NHL_reg_2016.head()

comp_id         competition_name        date   gid  goals_against  \
0     2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1     2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2     2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3     2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4     2734  2016 NHL Regular Season  2016-10-13  2738            7.0   

   goals_for  hgd home_away            team_name  tid  ... type  win  win.ord  \
0        5.0  1.0      home      Ottawa Senators   32  ...    2  1.0      2.0   
1        4.0  1.0      away  Toronto Maple Leafs    1  ...    2  0.0      0.0   
2        2.0 -3.0      home   Chicago Blackhawks    6  ...    2  0.0      0.0   
3        5.0 -3.0      away      St. Louis Blues    7  ...    2  1.0      2.0   
4        4.0  3.0      away       Calgary Flames   11  ...    2  0.0      0.0   

   year  away  home  cumGF  cumGA  pyth_wpct  fitted_results  
0  2016     0     1    5.0    4.0   0.609756               2  
1  2016     1     0    4.0    5.0   0.390244               0  
2  2016     0     1    2.0    5.0   0.137931               0  
3  2016     1     0    5.0    2.0   0.862069               2  
4  2016     1     0    4.0    7.0   0.246154               0  

[5 rows x 21 columns]

In [22]:
NHL_reg_2016['TRUE'] = np.where(NHL_reg_2016['fitted_results']==NHL_reg_2016['win.ord'], 1, 0)
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid  ...  win  win.ord  \
0           5.0  1.0      home      Ottawa Senators   32  ...  1.0      2.0   
1           4.0  1.0      away  Toronto Maple Leafs    1  ...  0.0      0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6  ...  0.0      0.0   
3           5.0 -3.0      away      St. Louis Blues    7  ...  1.0      2.0   
4           4.0  3.0      away       Calgary Flames   11  ...  0.0      0.0   
...         ...  ...       ...                  ...  ...  ...  ...      ...   
2419        0.0 -2.0      home  Washington Capitals    5  ...  0.0      0.0   
2420        3.0  1.0      away    Los Angeles Kings   41  ...  0.0      0.0   
2421        4.0  1.0      home        Anaheim Ducks   21  ...  1.0      2.0   
2422        2.0  3.0      away    Vancouver Canucks   53  ...  0.0      0.0   
2423        5.0  3.0      home      Edmonton Oilers   10  ...  1.0      2.0   

      year  away  home  cumGF  cumGA  pyth_wpct  fitted_results  TRUE  
0     2016     0     1    5.0    4.0   0.609756               2     1  
1     2016     1     0    4.0    5.0   0.390244               0     1  
2     2016     0     1    2.0    5.0   0.137931               0     1  
3     2016     1     0    5.0    2.0   0.862069               2     1  
4     2016     1     0    4.0    7.0   0.246154               0     1  
...    ...   ...   ...    ...    ...        ...             ...   ...  
2419  2016     0     1  265.0  186.0   0.669952               2     0  
2420  2016     1     0  197.0  204.0   0.482549               0     1  
2421  2016     0     1  233.0  210.0   0.551779               2     1  
2422  2016     1     0  178.0  238.0   0.358708               0     1  
2423  2016     0     1  245.0  211.0   0.574149               2     1  

[2424 rows x 22 columns]

In [23]:
# Obtain the success rate
Total = NHL_reg_2016['TRUE'].sum()
print(Total/2424)

0.6039603960396039


### 2.4) Obtain the fitted values by applying the ordered logit model directly
We obtained 1) linear product of the regression model, 2) fitted proabilities of each event, and 3) fitted game results from the probabilities using python commands above. Alternatively, we can also get all of these statistics by using the regression equation directly. Below you can find that the resulting columns are equavalient to the fitted values that we obtained above. 

In [24]:
# Obtain the fitted probabilities by using the regression coefficients
NHL_reg_2016['predL'] = 1/(1+np.exp(-(ol.coef_[1]-ol.coef_[0]*NHL_reg_2016['pyth_wpct'])))
NHL_reg_2016['predD'] = 1/(1+np.exp(-(ol.coef_[2]-ol.coef_[0]*NHL_reg_2016['pyth_wpct']))) - NHL_reg_2016['predL']
NHL_reg_2016['predW'] = 1- NHL_reg_2016['predL'] - NHL_reg_2016['predD']

# Display the outcome
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid  ... away  home  \
0           5.0  1.0      home      Ottawa Senators   32  ...    0     1   
1           4.0  1.0      away  Toronto Maple Leafs    1  ...    1     0   
2           2.0 -3.0      home   Chicago Blackhawks    6  ...    0     1   
3           5.0 -3.0      away      St. Louis Blues    7  ...    1     0   
4           4.0  3.0      away       Calgary Flames   11  ...    1     0   
...         ...  ...       ...                  ...  ...  ...  ...   ...   
2419        0.0 -2.0      home  Washington Capitals    5  ...    0     1   
2420        3.0  1.0      away    Los Angeles Kings   41  ...    1     0   
2421        4.0  1.0      home        Anaheim Ducks   21  ...    0     1   
2422        2.0  3.0      away    Vancouver Canucks   53  ...    1     0   
2423        5.0  3.0      home      Edmonton Oilers   10  ...    0     1   

      cumGF  cumGA  pyth_wpct  fitted_results  TRUE     predL     predD  \
0       5.0    4.0   0.609756               2     1  0.345124  0.008918   
1       4.0    5.0   0.390244               0     1  0.641852  0.008965   
2       2.0    5.0   0.137931               0     1  0.879770  0.004087   
3       5.0    2.0   0.862069               2     1  0.114317  0.004032   
4       4.0    7.0   0.246154               0     1  0.800085  0.006200   
...     ...    ...        ...             ...   ...       ...       ...   
2419  265.0  186.0   0.669952               2     0  0.273651  0.007865   
2420  197.0  204.0   0.482549               0     1  0.517178  0.009786   
2421  233.0  210.0   0.551779               2     1  0.421340  0.009591   
2422  178.0  238.0   0.358708               0     1  0.681190  0.008457   
2423  245.0  211.0   0.574149               2     1  0.391263  0.009380   

         predW  
0     0.645958  
1     0.349183  
2     0.116143  
3     0.881651  
4     0.193715  
...        ...  
2419  0.718484  
2420  0.473036  
2421  0.569069  
2422  0.310353  
2423  0.599356  

[2424 rows x 25 columns]

In [25]:
# Obtain the fitted game results by using the regression coefficients
NHL_reg_2016.loc[(NHL_reg_2016.predL > NHL_reg_2016.predW) & (NHL_reg_2016.predL > NHL_reg_2016.predD), 'fitted_result2'] = 0
NHL_reg_2016.loc[(NHL_reg_2016.predW > NHL_reg_2016.predL) & (NHL_reg_2016.predW > NHL_reg_2016.predD), 'fitted_result2'] = 2
NHL_reg_2016.loc[(NHL_reg_2016.predD > NHL_reg_2016.predW) & (NHL_reg_2016.predD > NHL_reg_2016.predL), 'fitted_result2'] = 1

pd.set_option('display.max_columns',30)

# Display the outcome
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
0           5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1           4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3           5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4           4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
...         ...  ...       ...                  ...  ...     ...   ...  ...   
2419        0.0 -2.0      home  Washington Capitals    5     WSH     2  0.0   
2420        3.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
2421        4.0  1.0      home        Anaheim Ducks   21     ANA     2  1.0   
2422        2.0  3.0      away    Vancouver Canucks   53     VAN     2  0.0   
2423        5.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   

      win.ord  year  away  home  cumGF  cumGA  pyth_wpct  fitted_results  \
0         2.0  2016     0     1    5.0    4.0   0.609756               2   
1         0.0  2016     1     0    4.0    5.0   0.390244               0   
2         0.0  2016     0     1    2.0    5.0   0.137931               0   
3         2.0  2016     1     0    5.0    2.0   0.862069               2   
4         0.0  2016     1     0    4.0    7.0   0.246154               0   
...       ...   ...   ...   ...    ...    ...        ...             ...   
2419      0.0  2016     0     1  265.0  186.0   0.669952               2   
2420      0.0  2016     1     0  197.0  204.0   0.482549               0   
2421      2.0  2016     0     1  233.0  210.0   0.551779               2   
2422      0.0  2016     1     0  178.0  238.0   0.358708               0   
2423      2.0  2016     0     1  245.0  211.0   0.574149               2   

      TRUE     predL     predD     predW  fitted_result2  
0        1  0.345124  0.008918  0.645958             2.0  
1        1  0.641852  0.008965  0.349183             0.0  
2        1  0.879770  0.004087  0.116143             0.0  
3        1  0.114317  0.004032  0.881651             2.0  
4        1  0.800085  0.006200  0.193715             0.0  
...    ...       ...       ...       ...             ...  
2419     0  0.273651  0.007865  0.718484             2.0  
2420     1  0.517178  0.009786  0.473036             0.0  
2421     1  0.421340  0.009591  0.569069             2.0  
2422     1  0.681190  0.008457  0.310353             0.0  
2423     1  0.391263  0.009380  0.599356             2.0  

[2424 rows x 26 columns]

### 2.5) Now we assess the accuracy of the ordit model in comparison to the true values

In [26]:
# Obtain the number of fitted values same as the actual outcomes
NHL_reg_2016['TRUE'] = np.where(NHL_reg_2016['fitted_result2']==NHL_reg_2016['win.ord'], 1, 0)
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
0           5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1           4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3           5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4           4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
...         ...  ...       ...                  ...  ...     ...   ...  ...   
2419        0.0 -2.0      home  Washington Capitals    5     WSH     2  0.0   
2420        3.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
2421        4.0  1.0      home        Anaheim Ducks   21     ANA     2  1.0   
2422        2.0  3.0      away    Vancouver Canucks   53     VAN     2  0.0   
2423        5.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   

      win.ord  year  away  home  cumGF  cumGA  pyth_wpct  fitted_results  \
0         2.0  2016     0     1    5.0    4.0   0.609756               2   
1         0.0  2016     1     0    4.0    5.0   0.390244               0   
2         0.0  2016     0     1    2.0    5.0   0.137931               0   
3         2.0  2016     1     0    5.0    2.0   0.862069               2   
4         0.0  2016     1     0    4.0    7.0   0.246154               0   
...       ...   ...   ...   ...    ...    ...        ...             ...   
2419      0.0  2016     0     1  265.0  186.0   0.669952               2   
2420      0.0  2016     1     0  197.0  204.0   0.482549               0   
2421      2.0  2016     0     1  233.0  210.0   0.551779               2   
2422      0.0  2016     1     0  178.0  238.0   0.358708               0   
2423      2.0  2016     0     1  245.0  211.0   0.574149               2   

      TRUE     predL     predD     predW  fitted_result2  
0        1  0.345124  0.008918  0.645958             2.0  
1        1  0.641852  0.008965  0.349183             0.0  
2        1  0.879770  0.004087  0.116143             0.0  
3        1  0.114317  0.004032  0.881651             2.0  
4        1  0.800085  0.006200  0.193715             0.0  
...    ...       ...       ...       ...             ...  
2419     0  0.273651  0.007865  0.718484             2.0  
2420     1  0.517178  0.009786  0.473036             0.0  
2421     1  0.421340  0.009591  0.569069             2.0  
2422     1  0.681190  0.008457  0.310353             0.0  
2423     1  0.391263  0.009380  0.599356             2.0  

[2424 rows x 26 columns]

In [27]:
# Obtain the success rate
Total = NHL_reg_2016['TRUE'].sum()
print(Total/2424)

0.6039603960396039


### 3) We can also incorporate the home team advantage into the ordered logit model above

In [28]:
iv = ['pyth_wpct', 'home']

In [29]:
ol = OrderedLogit()
ol.fit(NHL_reg_2016[iv], NHL_reg_2016['win.ord'])
ol.print_summary()

n=2424
                  beta  se(beta)      p  lower 0.95  upper 0.95     
attribute names                                                     
pyth_wpct       5.5906    0.4325 0.0000      4.7430      6.4383  ***
home            0.4849    0.0847 0.0000      0.3190      0.6509  ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Somers' D = 0.233


In [30]:
#%% To get the coefficients and the intercepts
print(f'pyth_wpct = {ol.coef_[0]:.4f}')
print(f'home = {ol.coef_[1]:.4f}')
print(f'interceptL|D = {ol.coef_[2]:.4f}')
print(f'interceptD|W = {ol.coef_[3]:.4f}')
print(ol.coef_)

pyth_wpct = 5.5906
home = 0.4849
interceptL|D = 3.0088
interceptD|W = 3.0486
[5.59063428 0.48494787 3.00882005 3.0485732 ]


In [31]:
NHL_reg_2016['predL2'] = 1/(1+np.exp(-(ol.coef_[2]-ol.coef_[0]*NHL_reg_2016['pyth_wpct']-ol.coef_[1]*NHL_reg_2016['home'])))
NHL_reg_2016['predD2'] = 1/(1+np.exp(-(ol.coef_[3]-ol.coef_[0]*NHL_reg_2016['pyth_wpct']-ol.coef_[1]*NHL_reg_2016['home']))) - NHL_reg_2016['predL']
NHL_reg_2016['predW2'] = 1- NHL_reg_2016['predL'] - NHL_reg_2016['predD']
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
0           5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1           4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3           5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4           4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
...         ...  ...       ...                  ...  ...     ...   ...  ...   
2419        0.0 -2.0      home  Washington Capitals    5     WSH     2  0.0   
2420        3.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
2421        4.0  1.0      home        Anaheim Ducks   21     ANA     2  1.0   
2422        2.0  3.0      away    Vancouver Canucks   53     VAN     2  0.0   
2423        5.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   

      win.ord  year  away  home  cumGF  cumGA  pyth_wpct  fitted_results  \
0         2.0  2016     0     1    5.0    4.0   0.609756               2   
1         0.0  2016     1     0    4.0    5.0   0.390244               0   
2         0.0  2016     0     1    2.0    5.0   0.137931               0   
3         2.0  2016     1     0    5.0    2.0   0.862069               2   
4         0.0  2016     1     0    4.0    7.0   0.246154               0   
...       ...   ...   ...   ...    ...    ...        ...             ...   
2419      0.0  2016     0     1  265.0  186.0   0.669952               2   
2420      0.0  2016     1     0  197.0  204.0   0.482549               0   
2421      2.0  2016     0     1  233.0  210.0   0.551779               2   
2422      0.0  2016     1     0  178.0  238.0   0.358708               0   
2423      2.0  2016     0     1  245.0  211.0   0.574149               2   

      TRUE     predL     predD     predW  fitted_result2    predL2    predD2  \
0        1  0.345124  0.008918  0.645958             2.0  0.292132 -0.044704   
1        1  0.641852  0.008965  0.349183             0.0  0.695743  0.062241   
2        1  0.879770  0.004087  0.116143             0.0  0.852299 -0.022536   
3        1  0.114317  0.004032  0.881651             2.0  0.140554  0.031108   
4        1  0.800085  0.006200  0.193715             0.0  0.836534  0.041813   
...    ...       ...       ...       ...             ...       ...       ...   
2419     0  0.273651  0.007865  0.718484             2.0  0.227658 -0.038927   
2420     1  0.517178  0.009786  0.473036             0.0  0.577145  0.069638   
2421     1  0.421340  0.009591  0.569069             2.0  0.363334 -0.048761   
2422     1  0.681190  0.008457  0.310353             0.0  0.731729  0.058270   
2423     1  0.391263  0.009380  0.599356             2.0  0.334927 -0.047424   

        predW2  
0     0.645958  
1     0.349183  
2     0.116143  
3     0.881651  
4     0.193715  
...        ...  
2419  0.718484  
2420  0.473036  
2421  0.569069  
2422  0.310353  
2423  0.599356  

[2424 rows x 29 columns]

In [32]:
NHL_reg_2016.loc[(NHL_reg_2016.predL2 > NHL_reg_2016.predW2) & (NHL_reg_2016.predL2 > NHL_reg_2016.predD2), 'fitted'] = 0
NHL_reg_2016.loc[(NHL_reg_2016.predW2 > NHL_reg_2016.predL2) & (NHL_reg_2016.predW2 > NHL_reg_2016.predD2), 'fitted'] = 2
NHL_reg_2016.loc[(NHL_reg_2016.predD2 > NHL_reg_2016.predW2) & (NHL_reg_2016.predD2 > NHL_reg_2016.predL2), 'fitted'] = 1

In [33]:
NHL_reg_2016['TRUE2'] = np.where(NHL_reg_2016['fitted']==NHL_reg_2016['win.ord'], 1, 0)
display(NHL_reg_2016)

comp_id         competition_name        date   gid  goals_against  \
0        2734  2016 NHL Regular Season  2016-10-12  2734            4.0   
1        2734  2016 NHL Regular Season  2016-10-12  2734            5.0   
2        2734  2016 NHL Regular Season  2016-10-13  2736            5.0   
3        2734  2016 NHL Regular Season  2016-10-13  2736            2.0   
4        2734  2016 NHL Regular Season  2016-10-13  2738            7.0   
...       ...                      ...         ...   ...            ...   
2419     2734  2016 NHL Regular Season  2017-04-09  5176            2.0   
2420     2734  2016 NHL Regular Season  2017-04-10  5177            4.0   
2421     2734  2016 NHL Regular Season  2017-04-10  5177            3.0   
2422     2734  2016 NHL Regular Season  2017-04-10  5179            5.0   
2423     2734  2016 NHL Regular Season  2017-04-10  5179            2.0   

      goals_for  hgd home_away            team_name  tid tricode  type  win  \
0           5.0  1.0      home      Ottawa Senators   32     OTT     2  1.0   
1           4.0  1.0      away  Toronto Maple Leafs    1     TOR     2  0.0   
2           2.0 -3.0      home   Chicago Blackhawks    6     CHI     2  0.0   
3           5.0 -3.0      away      St. Louis Blues    7     STL     2  1.0   
4           4.0  3.0      away       Calgary Flames   11     CGY     2  0.0   
...         ...  ...       ...                  ...  ...     ...   ...  ...   
2419        0.0 -2.0      home  Washington Capitals    5     WSH     2  0.0   
2420        3.0  1.0      away    Los Angeles Kings   41     LAK     2  0.0   
2421        4.0  1.0      home        Anaheim Ducks   21     ANA     2  1.0   
2422        2.0  3.0      away    Vancouver Canucks   53     VAN     2  0.0   
2423        5.0  3.0      home      Edmonton Oilers   10     EDM     2  1.0   

      win.ord  year  ...  home  cumGF  cumGA  pyth_wpct  fitted_results  TRUE  \
0         2.0  2016  ...     1    5.0    4.0   0.609756               2     1   
1         0.0  2016  ...     0    4.0    5.0   0.390244               0     1   
2         0.0  2016  ...     1    2.0    5.0   0.137931               0     1   
3         2.0  2016  ...     0    5.0    2.0   0.862069               2     1   
4         0.0  2016  ...     0    4.0    7.0   0.246154               0     1   
...       ...   ...  ...   ...    ...    ...        ...             ...   ...   
2419      0.0  2016  ...     1  265.0  186.0   0.669952               2     0   
2420      0.0  2016  ...     0  197.0  204.0   0.482549               0     1   
2421      2.0  2016  ...     1  233.0  210.0   0.551779               2     1   
2422      0.0  2016  ...     0  178.0  238.0   0.358708               0     1   
2423      2.0  2016  ...     1  245.0  211.0   0.574149               2     1   

         predL     predD     predW  fitted_result2    predL2    predD2  \
0     0.345124  0.008918  0.645958             2.0  0.292132 -0.044704   
1     0.641852  0.008965  0.349183             0.0  0.695743  0.062241   
2     0.879770  0.004087  0.116143             0.0  0.852299 -0.022536   
3     0.114317  0.004032  0.881651             2.0  0.140554  0.031108   
4     0.800085  0.006200  0.193715             0.0  0.836534  0.041813   
...        ...       ...       ...             ...       ...       ...   
2419  0.273651  0.007865  0.718484             2.0  0.227658 -0.038927   
2420  0.517178  0.009786  0.473036             0.0  0.577145  0.069638   
2421  0.421340  0.009591  0.569069             2.0  0.363334 -0.048761   
2422  0.681190  0.008457  0.310353             0.0  0.731729  0.058270   
2423  0.391263  0.009380  0.599356             2.0  0.334927 -0.047424   

        predW2  fitted  TRUE2  
0     0.645958     2.0      1  
1     0.349183     0.0      1  
2     0.116143     0.0      1  
3     0.881651     2.0      1  
4     0.193715     0.0      1  
...        ...     ...    ...  
2419  0.718484     2.0      0  
2420  0.473036     0.0      1  
2421  

In [34]:
Total = NHL_reg_2016['TRUE2'].sum()
print(Total/2424)

0.6126237623762376


## Conclusion

The accuracy rate of the model is about 61% in prediction, which is slightly better than the logit model. 